core/src/main/java/org/infinispan/marshall/core/internal/InternalExternalizerTable.java

```diff
@@ -0,0 +1,468 @@
+import org.infinispan.remoting.transport.jgroups.JGroupsAddress;
+import org.infinispan.remoting.transport.jgroups.JGroupsTopologyAwareAddress;
+import org.infinispan.statetransfer.StateChunk;
+import org.infinispan.statetransfer.TransactionInfo;
+import org.infinispan.stream.StreamMarshalling;
+import org.infinispan.stream.impl.intops.IntermediateOperationExternalizer;
+import org.infinispan.stream.impl.termop.TerminalOperationExternalizer;
+import org.infinispan.topology.CacheJoinInfo;
+import org.infinispan.topology.CacheStatusResponse;
+import org.infinispan.topology.CacheTopology;
+import org.infinispan.topology.ManagerStatusResponse;
+import org.infinispan.topology.PersistentUUID;
+import org.infinispan.transaction.xa.DldGlobalTransaction;
+import org.infinispan.transaction.xa.GlobalTransaction;
+import org.infinispan.transaction.xa.recovery.InDoubtTxInfoImpl;
+import org.infinispan.transaction.xa.recovery.RecoveryAwareDldGlobalTransaction;
+import org.infinispan.transaction.xa.recovery.RecoveryAwareGlobalTransaction;
+import org.infinispan.transaction.xa.recovery.SerializableXid;
+import org.infinispan.util.IntObjectHashMap;
+import org.infinispan.util.IntObjectMap;
+import org.infinispan.util.KeyValuePair;
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+import org.infinispan.xsite.statetransfer.XSiteState;
+import org.jboss.marshalling.util.IdentityIntMap;
+
+final class InternalExternalizerTable {
+
+   private static final Log log = LogFactory.getLog(InternalMarshaller.class);
+   private static final boolean trace = log.isTraceEnabled();
+
+   private static final int NOT_FOUND = -1;
+
+   /**
+    * Encoding of the internal marshaller
+    */
+   private final Encoding enc;
+
+   /**
+    * Pairing between internally marshallable types and their numeric identifiers.
+    * Externalizer identifiers are decided based on order of internal externalizer
+    * definition below. Such limitation speeds up internal externalizer lookups
+    * at read time using a simple array of externalizers.
+    *
+    * This collection is sized according to an approximation of number of
+    * internal types.
+    */
+   final IdentityIntMap<Class<?>> internalExtIds = new IdentityIntMap<>(320);
+
+   /**
+    * Array of internal externalizers used for externalizer lookups on read time.
+    * Since each externalizer's id is within a sequence of numbers decided on
+    * startup, lookups can be done via an array making it very efficient.
+    *
+    * This collection is sized according to an approximation of number of
+    * internal externalizers.
+    */
+   final AdvancedExternalizer[] internalExts = new AdvancedExternalizer[128];
+
+   /**
+    * Pairing between types marshalled by the externalizers predefined via
+    * global configuration and their externalizer identifiers.
+    */
+   final IdentityIntMap<Class<?>> predefExtIds = new IdentityIntMap<>(32);
```

Comment:
I have another opinion here. Since the user-defined externalizers share the same interfaces than ours internal externalizers, they could be merge in the same data structure. We would end up with a ```Map<Class, AdvancedExternalizer>``` and an ```IntObjectHashMap<AdvancedExternalizer>```.
The ```InternalIds.INTERNAL``` and use and ```InternalIds.PREDEFINED``` could be merged.